@brainelectronics this is an amazing feature!! :partying_face: 

> * Example callback usage shown in [TCP client example](examples/tcp_client_example.py)

This url is broken. 

> The updated documentation can be found at the [RTD MicroPython Modbus optional callbacks section](https://micropython-modbus--51.org.readthedocs.build/en/51/USAGE.html#optional-callbacks)

That doc is very good! I used that as reference!

**Results:** works, with just one problem: using a list of `COILS,` after set it to new values, read it cause `OSError: invalid response CRC`, but the `Slave` `COIL` get callback function works. That problem do not happen on the `HREGS.` Details below:


**Callback functions and register_definitions:**

```Python
def my_coil_set_cb(reg_type, address, val):
    print('Custom callback, called on setting {} at {} to: {}'.
          format(reg_type, address, val))

def my_coil_get_cb(reg_type, address, val):
    print('Custom callback, called on getting {} at {}, currently: {}'.
          format(reg_type, address, val))

def my_hregs_set_cb(reg_type, address, val):
    print('Custom callback, called on setting {} at {} to: {}'.
          format(reg_type, address, val))

def my_hregs_get_cb(reg_type, address, val):
    print('Custom callback, called on getting {} at {}, currently: {}'.
          format(reg_type, address, val))

# common slave register setup, to be used with the Master example above
register_definitions = {
    "COILS": {
        "EXAMPLE_COIL": {
            "register": 123,
            "len": 1,
            "val": 1
        },
        "COIL_SIGNALS": {
            "register": 125,
            "len": 16,
            "val": [1,1,1,1,1,0,0,0,0,1,1,1,0,0,0,1],
            "on_get_cb": my_coil_get_cb,
            "on_set_cb": my_coil_set_cb
        }
    },
    "HREGS": {
        "EXAMPLE_HREG": {
            "register": 93,
            "len": 1,
            "val": 19
        },
        "HREG_VALUES": {
            "register": 130,
            "len": 33,
            "val": [34, 12, 14, 0, 10, 4, 2, 1345, 34, 8, 1100, 350, 456, 754, 324, 423, 530, 90, 320, 34, 244, 355, 606, 656, 640, 620, 677, 623, 234, 567, 34, 56, 68],
            "on_get_cb": my_hregs_get_cb,
            "on_set_cb": my_hregs_set_cb
        }
    },
    "ISTS": {
        "EXAMPLE_ISTS": {
            "register": 67,
            "len": 1,
            "val": 0
        }
    },
    "IREGS": {
        "EXAMPLE_IREG": {
            "register": 10,
            "len": 2,
            "val": 60001
        }
    }
}
```

**Slave RTU:**

```Python
$ mpremote run rtu_client_example_with_callback.py 
Running ModBus version: 2.3.0-rc22.dev51

Custom callback, called on getting COILS at 125, currently: [1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1]
Custom callback, called on setting COILS at 125 to: [False, True, True]
Custom callback, called on getting COILS at 125, currently: [False, True, True]
Custom callback, called on getting HREGS at 130, currently: [34, 12, 14, 0, 10, 4, 2, 1345, 34, 8, 1100, 350, 456, 754, 324, 423, 530, 90, 320, 34, 244, 355, 606, 656, 640, 620, 677, 623, 234, 567, 34, 56, 68]
Custom callback, called on setting HREGS at 130 to: [2, 65532, 6, 65280, 1024]
Custom callback, called on getting HREGS at 130, currently: [2, 65532, 6, 65280, 1024]
```

**Master RTU:**

```Python
>>> from umodbus import version
>>> print('Running ModBus version: {}'.format(version.__version__))
Running ModBus version: 2.3.0-rc22.dev51
>>> 
>>> from umodbus.serial import Serial as ModbusRTUMaster
>>> rtu_pins = (17, 18)
>>> 
>>> host = ModbusRTUMaster(baudrate=115200, data_bits=8, stop_bits=1, parity=None, pins=rtu_pins, ctrl_pin=15)
>>> 
>>> host.read_coils(slave_addr=10, starting_addr=125, coil_qty=16)
[True, True, True, True, True, False, False, False, False, True, True, True, False, False, False, True]
>>> host.write_multiple_coils(slave_addr=10, starting_address=125, output_values=[1,1,0])
True
>>> host.read_coils(slave_addr=10, starting_addr=125, coil_qty=16) <--- Here happen the error, after set to new values
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/lib/umodbus/common.py", line 136, in read_coils
  File "/lib/umodbus/serial.py", line 289, in _send_receive
  File "/lib/umodbus/serial.py", line 322, in _validate_resp_hdr
OSError: invalid response CRC
>>> host.read_holding_registers(slave_addr=10, starting_addr=130, register_qty=33, signed=False)
(34, 12, 14, 0, 10, 4, 2, 1345, 34, 8, 1100, 350, 456, 754, 324, 423, 530, 90, 320, 34, 244, 355, 606, 656, 640, 620, 677, 623, 234, 567, 34, 56, 68)
>>> host.write_multiple_registers(slave_addr=10, starting_address=130, register_values=[2, -4, 6, -256, 1024], signed=True)
True
>>> host.read_holding_registers(slave_addr=10, starting_addr=130, register_qty=33, signed=False)
(2, 65532, 6, 65280, 1024)
>>> 
```

--- a/docs/USAGE.md
+++ b/docs/USAGE.md
@@ -231,8 +231,9 @@ be `percent`.
 ###### Optional callbacks
 
 The optional keys `on_set_cb` and `on_get_cb` can be used to register a
-callback function on client side which is executed after a new value has been
-set or after a register value has been requested.
+callback function on client side which is executed **after** a new value has
+been set or **before** the response of a requested register value has been
+sent.
 
 ```{note}
 Getter callbacks can be registered for all registers with the `on_get_cb`

--- a/examples/tcp_client_example.py
+++ b/examples/tcp_client_example.py
@@ -107,7 +107,7 @@ def my_inputs_register_get_cb(reg_type, address, val):
 
     # any operation should be as short as possible to avoid response timeouts
     # It would be also possible to read the leatest pin state at this time
-    val += 1
+    val[0] += 1
     client.set_ist(address=address, value=val)
     print('Incremented current value by +1 before sending response')
 

--- a/examples/tcp_client_example.py
+++ b/examples/tcp_client_example.py
@@ -99,8 +99,17 @@ def my_discrete_inputs_register_get_cb(reg_type, address, val):
 
 
 def my_inputs_register_get_cb(reg_type, address, val):
+    # usage of global isn't great, but okay for an example
+    global client
+
     print('Custom callback, called on getting {} at {}, currently: {}'.
           format(reg_type, address, val))
+
+    # any operation should be as short as possible to avoid response timeouts
+    # It would be also possible to read the leatest pin state at this time
+    val += 1
+    client.set_ist(address=address, value=val)
+    print('Incremented current value by +1 before sending response')
 
 
 def reset_data_registers_cb(reg_type, address, val):

--- a/tests/test_tcp_example.py
+++ b/tests/test_tcp_example.py
@@ -542,6 +542,10 @@ class TestTcpExample(unittest.TestCase):
         expectation = \
             (self._register_definitions['IREGS']['EXAMPLE_IREG']['val'], )
 
+        # due to value increment by registered callback in
+        # tcp_client_example.py, see #31 and #51
+        expectation[0] += 1
+
         register_value = self._host.read_input_registers(
             slave_addr=self._client_addr,
             starting_addr=ireg_address,

--- a/umodbus/modbus.py
+++ b/umodbus/modbus.py
@@ -140,12 +140,15 @@ class Modbus(object):
         address = request.register_addr
 
         if address in self._register_dict[reg_type]:
-            vals = self._create_response(request=request, reg_type=reg_type)
-            request.send_response(vals)
 
             if self._register_dict[reg_type][address].get('on_get_cb', 0):
+                vals = self._create_response(request=request,
+                                             reg_type=reg_type)
                 _cb = self._register_dict[reg_type][address]['on_get_cb']
                 _cb(reg_type=reg_type, address=address, val=vals)
+
+            vals = self._create_response(request=request, reg_type=reg_type)
+            request.send_response(vals)
         else:
             request.send_exception(Const.ILLEGAL_DATA_ADDRESS)
 
@@ -382,7 +385,7 @@ class Modbus(object):
 
     def remove_ist(self, address: int) -> Union[None, bool, List[bool]]:
         """
-        Remove a holding register from the modbus register dictionary.
+        Remove a discrete input register from the modbus register dictionary.
 
         :param      address:  The address (ID) of the register
         :type       address:  int

